### ISSUE TITLE
stack failed to install ghc but broke my backups

### ISSUE THREAD

**Issue by pushcx** [commercialhaskell/stack#581]
I'm a [newbie Haskell programmer](https://plus.google.com/+MichaelSnoyman/posts/BEqbGbjiBpa). I removed my Arch Linux install of Haskell and installed Slack from the AUR. Tinkering around to get helloworld.hs running in ghci or as a binary, I followed the README.

1. `stack build` told me to run `stack new`.
2. `stack new` vomited a tree of files all over the current directory (thanks?), then downloaded something. It didn't suggest any next step.
3. I ran `stack build` again and it suggested `stack setup`.
4. `stack setup` claimed to download and install GHC, but `ghc` and `ghci` just return "command not found". It did say "Would add the following to PATH:" but it seems nuts to have to edit my PATH for every single project I have, so I don't know what to do with that.
5. `stack build` again? It compiled new-template. Still no `ghc` or `ghci` commands.
6. None of the files `stack new` left for me provide any hints. The README suggests running `stack path` and `stack`, The [wiki](https://github.com/commercialhaskell/stack/wiki) pointed me back to the README, but none of the other pages explain why ghc didn't get installed.
7. `stack build` now does nothing and says nothing.
8. Backing up my home directory failed (**!**) because there's a new 1.4GB of files in ~ hiding in .slack. The directory mixes a config file I'd want to back up (~/.slack/global/stack.yaml) with caching, packaging, and binaries that I'd never want to back up. Nothing in the wiki or man page explains how to get this garbage out of my dotfiles.
9. Looking for more info, I learn the word "stack" is ungoogleable. Did y'all know that "stack" is already a programming term? I give up.

I suppose I have four issues:
1. Somewhere in here there's a missing step required for installing ghc.
2. There's no documentation of the files `stack new` left in my working directory.
3. Installing `stack` does not magically increase my bandwidth so much that I don't notice an extra 1.4GB prentending to be dotfiles. Breaking my backups is a huge problem.
4. "stack" is not a searchable name.

**Comment by mgsloan** [comment]
Thanks for the detail and explanation, you've pointed out some interesting usability issues.

> stack setup claimed to download and install GHC, but ghc and ghci just return "command not found". It did say "Would add the following to PATH:" but it seems nuts to have to edit my PATH for every single project I have, so I don't know what to do with that.

`stack setup` only needs to be run when you need a particular ghc version installed, so you only need to add to your PATH then.  Adding this to your PATH is likely the missing step.

**Comment by pushcx** [comment]
Thanks for your response. I started fresh (removing ~/.stack and the many files left by `stack new`) and adding the path `stack setup` gave me did leave me a working `ghc` and `ghci`.

It looks like that "Would add the following" message comes from [Main.hs, lines 444-445](https://github.com/commercialhaskell/stack/blob/b14e960449b28dabd79e96490b3e9e2c62c43b8a/src/main/Main.hs#L444):

```
                  Nothing -> $logInfo "GHC on PATH would be used"
                  Just ps -> $logInfo $ "Would add the following to PATH: "
```

I can't follow all of what `ensureGHC` does; I think it returns the path to GHC if installed or installs it if not.

I think changing the second message to say something like "Add this directory to your PATH for this version of GHC: " would explain what the next step is and that it's not a per-project setting. I'm not opening a PR because I can't read the code well enough to be sure it wouldn't just be a waste. My concern is that there are two booleans here: whether GHC is installed, and whether that install on the current PATH, and the messages need to reflect the three possible outcomes.

(I suppose folks unfamiliar with the shell would be lost by the instruction. This is the best message I can think of at the moment, maybe there's a standard "how to add things to your PATH" doc that could be linked to, or maybe the project is comfortable requiring that familiarity with shell setup.)

Thanks, @mgsloan, for solving my first problem. I hope this helps other people avoid it in the future and I'm happy to contribute back a bit. :)

**Comment by snoyberg** [comment]
I think the core issue here is an assumption that `stack setup` would do something global, which it doesn't. I'd like to fix this at the documentation level. Can you clarify what implied that `setup` was a global installation command?

For doing things with `ghc` directly, the recommended approach is to use `stack ghc`, not to modify the `PATH` for your user. You can do be, but it's not the intended use case, which is why the output for `stack setup` doesn't tell you to do so.

As for putting things under ~/.stack: that's just the way stack works, and is in fact [the first FAQ entry](https://github.com/commercialhaskell/stack/wiki/FAQ#where-is-stack-installed-and-will-it-interfere-with-ghc-etc-i-already-have-installed). Even if GHC itself wasn't installed in your home directory, the binary versions of your packages would be, which will add up to quite a bit of space. This isn't a stack-unique problem; my ~/.cabal directory previously has easily crossed 4GB.

Improved documentation seems like the right approach here, so if you have any suggestions on where things could be clarified, please do indicate so. But there are no changes to be made to how stack functions that I can see from this issue, so closing.

**Comment by pushcx** [comment]
I thought "setup" was a global installation command because nothing implied it wasn't. There were no pointers to run `stack ghc` and it's not documented in the readme or wiki.

Can ~/.stack be safely symlinked so it's not mixed in with dotfiles? That'd be a good note to have in the FAQ. Cabal dumping binaries in personal files is not a relevant reason for stack breaking backups, but at least there'd be a fix. I'd rather leave the undocumented global config file mixed in there not backed up than back up nothing in my home directory.

There were two other issues unaddressed here: there's no docs anywhere for the files left by `stack new` and the name 'stack' is not searchable.

**Comment by snoyberg** [comment]
After discussing with @bitemyapp, we're creating a new "Support" milestone instead of redirecting to the mailing list or similar for these kinds of issues. Reopening.

**Comment by snoyberg** [comment]
Actually, rereading this issue, I see a confused some of the points above about who said what. Initial closing was *not* the right call, my apologies @pushcx.

**Comment by snoyberg** [comment]
I've pushed a commit to change the final output of `stack setup` to inform about the relevant commands:

```
$ stack setup
Would add the following to PATH: /home/vagrant/.stack/programs/x86_64-linux/ghc-7.8.4/bin
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
```

Regarding the other points

* [Naming open discussion](https://www.reddit.com/r/haskell/comments/3d9z57/about_the_name_stack/)
* `stack new` is being rewritten as we speak, the current implementation was never meant as more than a stop-gap measure.

**Comment by bitemyapp** [comment]
@snoyberg I added a note to the wiki where `stack setup` is mentioned as well.

https://github.com/commercialhaskell/stack/wiki/FAQ#i-already-have-ghc-installed-can-i-still-use-stack

> I already have GHC installed, can I still use stack?

> Yes. stack will default to using whatever GHC is on your PATH. If that GHC is a compatible version with the snapshot you're using, it will simply use it. Otherwise, it will prompt you to run stack setup. Note that stack setup installs GHC into ~/.stack/programs/$platform/ghc-$version/ and not a global location.

> Note that GHC installation doesn't work for all OSes, so in some cases the first option will need to install GHC yourself.

**Comment by radix** [comment]
I literally don't understand what "Would add the following to PATH" means. Is it a suggestion that the user should add it to their PATH? It sounds like "[stack] would add the following to PATH, [if it weren't for X...]", which doesn't make sense to me either.

**Comment by snoyberg** [comment]
What it means is "if stack was going to actually do anything useful right now besides just making sure GHC is available, it would modify the PATH as follows before running subcommands." Yes, I agree that on looking again that's not clear from the message ;). Any suggestions? Perhaps:

> stack's modified PATH for subcommands would include ...

**Comment by radix** [comment]
@snoyberg I still don't think I understand. what do you mean by "do anything useful"? does it sometimes set that PATH and sometimes not?

Maybe it should just say "GHC has been installed in {path}\nYou may consider adding it to your PATH environment variable, or invoke ghc with `stack ghc` [etc]" or something? I don't think a "I would do this, but..." message is very useful/actionable to a user

**Comment by snoyberg** [comment]
Hmm... let me try again. If you were running a command like `stack build` that actually used GHC or other tools, then stack would use the version of GHC found on the given modified PATH. But it's slightly more general than that, since it also is used on Windows to pull in things like MSYS (used for `git`, `sh`, etc). The intention here was to let users know which GHC and other tools are being used, *not* to encourage them to do anything concrete with that information.

Maybe with the presence of `stack path`, the right thing to do is just say either:

* stack will use the GHC on your PATH
* stack will used a locally installed GHC

And then mention the commands `stack path`, `stack exec env`, or others as a "for more information" type deal.

**Comment by radix** [comment]
Ahh! Okay, I get it now. I think your new suggestion makes sense :)

**Comment by drwebb** [comment]
+1. The new suggestion seems nice to me.

**Comment by snoyberg** [comment]
Alright, wording updated (apparently my previous commit didn't push, so it's all in one commit).

**Comment by pushcx** [comment]
That's really helpful writing and totally addresses the confusion I had about `ghc` not working. I can't see it again for the first time, but I think it's solved.

Also, thanks for reopening this issue. I'm sorry I was rude; I was really grumpy about the broken backups.

Thinking about that part some more, does Haskell code often follow the [XDG spec](http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html)? I found [a package](https://hackage.haskell.org/package/xdg-basedir) with some support, not sure what else is out there. If the indices, downloads, and build dirs to ~/.cache/ghc they'd be in a directory that's explicitly for "user-specific non-essential (cached) data" that, if deleted, can be restored or regenerated trivially. There also seem to be folks talking about using .local/bin as a standard target, but I'm not aware of public definition (anyone here with more up-to-date with unix packaging trends?).

I'm told there's already discussions about the name "stack" happening elsewhere, so that can get ignored in this issue.

**Comment by snoyberg** [comment]
I'm open to discussing using the cache directory, but we should probably do so in its own issue. I'd like to understand better whether it's common for the cache files to be pruned, which probably isn't what people want.

**Comment by drwebb** [comment]
The git version of stack has the "local-bin-path" cli option and stack.yaml
option for changing where the binaries are placed.

**Comment by snoyberg** [comment]
Unless I'm mistaken, at this point we've covered the pertinent points, and `stack new` is still being worked on in a separate issue. Closing now, if there's something else to discuss, it's probably worth doing so in a separate issue.
